# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.

In [173]:
import pandas as pd

df_customer = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv')

display(df_customer.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Unnamed: 0                     10910 non-null  int64  
 1   Customer                       10910 non-null  object 
 2   State                          10279 non-null  object 
 3   Customer Lifetime Value        10910 non-null  float64
 4   Response                       10279 non-null  object 
 5   Coverage                       10910 non-null  object 
 6   Education                      10910 non-null  object 
 7   Effective To Date              10910 non-null  object 
 8   EmploymentStatus               10910 non-null  object 
 9   Gender                         10910 non-null  object 
 10  Income                         10910 non-null  int64  
 11  Location Code                  10910 non-null  object 
 12  Marital Status                 10910 non-null 

None

In [174]:
df_customer.head(3)

Unnamed: 0 Customer       State  Customer Lifetime Value Response Coverage  \
0           0  DK49336     Arizona              4809.216960       No    Basic   
1           1  KX64629  California              2228.525238       No    Basic   
2           2  LZ68649  Washington             14947.917300       No    Basic   

  Education Effective To Date EmploymentStatus Gender  ...  \
0   College           2/18/11         Employed      M  ...   
1   College           1/18/11       Unemployed      F  ...   
2  Bachelor           2/10/11         Employed      M  ...   

   Number of Open Complaints Number of Policies     Policy Type        Policy  \
0                        0.0                  9  Corporate Auto  Corporate L3   
1                        0.0                  1   Personal Auto   Personal L3   
2                        0.0                  2   Personal Auto   Personal L3   

   Renew Offer Type  Sales Channel  Total Claim Amount  Vehicle Class  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   

  Vehicle Size Vehicle Type  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  

[3 rows x 26 columns]

In [175]:
df_customer_claim = df_customer[df_customer['Response'] == 'Yes'] 
df_customer_claim = df_customer[df_customer['Total Claim Amount'] > 1000]

df_customer_claim.info()


<class 'pandas.core.frame.DataFrame'>
Index: 491 entries, 30 to 10909
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Unnamed: 0                     491 non-null    int64  
 1   Customer                       491 non-null    object 
 2   State                          466 non-null    object 
 3   Customer Lifetime Value        491 non-null    float64
 4   Response                       466 non-null    object 
 5   Coverage                       491 non-null    object 
 6   Education                      491 non-null    object 
 7   Effective To Date              491 non-null    object 
 8   EmploymentStatus               491 non-null    object 
 9   Gender                         491 non-null    object 
 10  Income                         491 non-null    int64  
 11  Location Code                  491 non-null    object 
 12  Marital Status                 491 non-null    objec

2. Using the original Dataframe, analyze the average total_claim_amount by each policy type and gender for customers who have responded "Yes" to the last marketing campaign. Write your conclusions.

In [176]:
df_customer_average = df_customer[df_customer['Response'] == 'Yes'].groupby(['Policy Type', 'Gender'])['Total Claim Amount'].mean()

df_customer_average

Policy Type     Gender
Corporate Auto  F         433.738499
                M         408.582459
Personal Auto   F         452.965929
                M         457.010178
Special Auto    F         453.280164
                M         429.527942
Name: Total Claim Amount, dtype: float64

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [177]:
df_customer_policy_count = df_customer.groupby('State')['Policy'].count() 
df_customer_policy_count = df_customer_policy_count[df_customer_policy_count > 500]

display(df_customer_policy_count)


State
Arizona       1937
California    3552
Nevada         993
Oregon        2909
Washington     888
Name: Policy, dtype: int64

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [178]:
df_customer.head(3)

Unnamed: 0 Customer       State  Customer Lifetime Value Response Coverage  \
0           0  DK49336     Arizona              4809.216960       No    Basic   
1           1  KX64629  California              2228.525238       No    Basic   
2           2  LZ68649  Washington             14947.917300       No    Basic   

  Education Effective To Date EmploymentStatus Gender  ...  \
0   College           2/18/11         Employed      M  ...   
1   College           1/18/11       Unemployed      F  ...   
2  Bachelor           2/10/11         Employed      M  ...   

   Number of Open Complaints Number of Policies     Policy Type        Policy  \
0                        0.0                  9  Corporate Auto  Corporate L3   
1                        0.0                  1   Personal Auto   Personal L3   
2                        0.0                  2   Personal Auto   Personal L3   

   Renew Offer Type  Sales Channel  Total Claim Amount  Vehicle Class  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   

  Vehicle Size Vehicle Type  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  

[3 rows x 26 columns]

In [179]:

df_customer_lifetime = df_customer.groupby(['Education', 'Gender'])['Customer Lifetime Value'].agg(['max', 'min', 'median'])
display(df_customer_lifetime) 

max          min       median
Education            Gender                                       
Bachelor             F       73225.95652  1904.000852  5640.505303
                     M       67907.27050  1898.007675  5548.031892
College              F       61850.18803  1898.683686  5623.611187
                     M       61134.68307  1918.119700  6005.847375
Doctor               F       44856.11397  2395.570000  5332.462694
                     M       32677.34284  2267.604038  5577.669457
High School or Below F       55277.44589  2144.921535  6039.553187
                     M       83325.38119  1940.981221  6286.731006
Master               F       51016.06704  2417.777032  5729.855012
                     M       50568.25912  2272.307310  5579.099207

## Conclusions from the Data:
### Gender Differences in Customer Lifetime Value (CLV):

In general, females tend to have a higher minimum CLV across most education levels compared to males. For example, females with a Bachelor's degree have a minimum CLV of 1904.00, slightly higher than males at 1898.00.
However, males tend to have higher maximum CLV at lower education levels. In the High School or Below category, males have a maximum CLV of 83325.38, which is significantly higher than females at 55277.45.
The median CLV does not show a consistent pattern by gender, with some categories showing higher medians for females and others for males. For example, males with High School or Below education have a median CLV of 6286.73, which is higher than females at 6039.55. On the other hand, females with a Master's degree have a median CLV of 5729.86, higher than males at 5579.10.

### Impact of Education Level on CLV:

Customers with lower education levels (e.g., High School or Below) show a wider range in CLV, particularly for males, where the maximum CLV is much higher than the minimum. This suggests that lower education levels may be associated with a higher variation in customer value, especially among males.
Conversely, customers with higher education levels (e.g., Doctor or Master) tend to have lower maximum CLV and less variation between the maximum and minimum CLV. For example, males with a Doctorate have a maximum CLV of 32677.34, which is much lower compared to those with lower education levels.

#### - High School Education and Gender:

For the High School or Below education level, males have the highest maximum CLV at 83325.38, indicating that some male customers with lower education levels can still be very valuable.
On the other hand, females in this category have a lower maximum CLV at 55277.45 but a more stable minimum and median CLV, suggesting more consistent value among female customers with lower education.
#### - Doctorate Holders Have Lower CLV:

Interestingly, customers with a Doctorate tend to have lower maximum and median CLV compared to other education levels. This could indicate that having a higher education level doesn't necessarily correlate with a higher lifetime value in this dataset.

### Summary:
The data suggests that both gender and education level influence customer lifetime value, but the impact is mixed. While males generally exhibit higher maximum CLVs at lower education levels, females show more stability with less variation in CLV across education levels. Higher education levels do not always correspond to higher customer lifetime values, and the relationship between education, gender, and customer value is complex and varies across categories.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

In [180]:
df_customer.head(3)

Unnamed: 0 Customer       State  Customer Lifetime Value Response Coverage  \
0           0  DK49336     Arizona              4809.216960       No    Basic   
1           1  KX64629  California              2228.525238       No    Basic   
2           2  LZ68649  Washington             14947.917300       No    Basic   

  Education Effective To Date EmploymentStatus Gender  ...  \
0   College           2/18/11         Employed      M  ...   
1   College           1/18/11       Unemployed      F  ...   
2  Bachelor           2/10/11         Employed      M  ...   

   Number of Open Complaints Number of Policies     Policy Type        Policy  \
0                        0.0                  9  Corporate Auto  Corporate L3   
1                        0.0                  1   Personal Auto   Personal L3   
2                        0.0                  2   Personal Auto   Personal L3   

   Renew Offer Type  Sales Channel  Total Claim Amount  Vehicle Class  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   

  Vehicle Size Vehicle Type  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  

[3 rows x 26 columns]

In [181]:
df_customer['Effective To Date'] = pd.to_datetime(df_customer['Effective To Date'])
df_customer.head(3)

C:\Users\eslab\AppData\Local\Temp\ipykernel_12936\3221681098.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df_customer['Effective To Date'] = pd.to_datetime(df_customer['Effective To Date'])


Unnamed: 0 Customer       State  Customer Lifetime Value Response Coverage  \
0           0  DK49336     Arizona              4809.216960       No    Basic   
1           1  KX64629  California              2228.525238       No    Basic   
2           2  LZ68649  Washington             14947.917300       No    Basic   

  Education Effective To Date EmploymentStatus Gender  ...  \
0   College        2011-02-18         Employed      M  ...   
1   College        2011-01-18       Unemployed      F  ...   
2  Bachelor        2011-02-10         Employed      M  ...   

   Number of Open Complaints Number of Policies     Policy Type        Policy  \
0                        0.0                  9  Corporate Auto  Corporate L3   
1                        0.0                  1   Personal Auto   Personal L3   
2                        0.0                  2   Personal Auto   Personal L3   

   Renew Offer Type  Sales Channel  Total Claim Amount  Vehicle Class  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   

  Vehicle Size Vehicle Type  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  

[3 rows x 26 columns]

In [191]:

df_customer_policies_sold = df_customer.groupby(['State', df_customer['Effective To Date'].dt.month])['Policy'].count()
df_customer_policies_sold

State       Effective To Date
Arizona     1                    1008
            2                     929
California  1                    1918
            2                    1634
Nevada      1                     551
            2                     442
Oregon      1                    1565
            2                    1344
Washington  1                     463
            2                     425
Name: Policy, dtype: int64

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

In [194]:
df_customer_policies_sold = df_customer_policies_sold.groupby('State').sum().sort_values(ascending=False).head(3)
df_customer_policies_sold

State
California    3552
Oregon        2909
Arizona       1937
Name: Policy, dtype: int64

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

In [200]:
display(df_customer.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   Unnamed: 0                     10910 non-null  int64         
 1   Customer                       10910 non-null  object        
 2   State                          10279 non-null  object        
 3   Customer Lifetime Value        10910 non-null  float64       
 4   Response                       10279 non-null  object        
 5   Coverage                       10910 non-null  object        
 6   Education                      10910 non-null  object        
 7   Effective To Date              10910 non-null  datetime64[ns]
 8   EmploymentStatus               10910 non-null  object        
 9   Gender                         10910 non-null  object        
 10  Income                         10910 non-null  int64         
 11  Location Code  

None

In [202]:
df_customer_melted = df_customer['Sales Channel'], df_customer[ 'Response']

KeyError: 'Response'

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9